File: pkg/IP/ipnet.go
package IP

import "strconv"

type Net struct {
	NetworkAddress IP   `json:"network_address"`
	CIDR           CIDR `json:"cidr"`
}

func newIPNet(ip IP, cidr CIDR) Net {
	return Net{
		NetworkAddress: IP{
			// calculate net address
			Value: ip.Value & cidr.Mask,
		},
		CIDR: cidr,
	}
}

func NewIPNetFromStr(ipStr string, cidrStr string) (Net, error) {
	ip, err := newIP(ipStr)
	if err != nil {
		return Net{}, err
	}
	cidr, err := NewCIDRFromString(cidrStr)
	if err != nil {
		return Net{}, err
	}
	return newIPNet(ip, cidr), nil
}

func NewIPNetFromMixed(ipStr string, cidrStr int) (Net, error) {
	ip, err := newIP(ipStr)
	if err != nil {
		return Net{}, err
	}
	cidr, err := NewCIDR(cidrStr)
	if err != nil {
		return Net{}, err
	}
	return newIPNet(ip, cidr), nil
}

func (net1 Net) ToString() string {
	return net1.NetworkAddress.toString() + "/" + strconv.Itoa(int(net1.GetRawCIDR()))
}

func (net1 Net) GetRawCIDR() uint8 {
	return net1.CIDR.Value
}

func (net1 Net) IsSubnetOf(net2 Net) bool {
	return net2.includesIP(net1.NetworkAddress) && net1.CIDR.Value >= net2.CIDR.Value
}

func (net1 Net) includesIP(ip IP) bool {
	return ip.Value&net1.CIDR.Mask == net1.NetworkAddress.Value
}

func (net1 Net) IsIdentical(net2 Net) bool {
	return net1.NetworkAddress.Value == net2.NetworkAddress.Value && net1.CIDR.Value == net2.CIDR.Value
}

File: pkg/IP/ipnet.py
from dataclasses import dataclass

from pkg.IP.ip import IP
from pkg.IP.cidr import CIDR

@dataclass
class Net:
    network_address: IP
    cidr: CIDR

    @classmethod
    def new_ip_net(cls, ip: IP, cidr: CIDR) -> 'Net':
        return cls(
            network_address=IP(
                # calculate net address
                value=ip.value & cidr.mask
            ),
            cidr=cidr
        )

    @classmethod
    def new_from_str(cls, ip_str: str, cidr_str: str) -> 'Net':
        ip = IP.new_from_string(ip_str)

        cidr = CIDR.new_from_string(cidr_str)

        return cls.new_ip_net(ip, cidr)

    @classmethod
    def new_from_mixed(cls, ip_str: str, cidr_int: int) -> 'Net':
        ip = IP.new_from_string(ip_str)

        cidr = CIDR.new_from_int(cidr_int)

        return cls.new_ip_net(ip, cidr)

    def to_string(self) -> str:
        return f"{self.network_address.to_string()}/{str(self.get_raw_cidr())}"

    def get_raw_cidr(self) -> int:
        return self.cidr.value

    def is_subnet_of(self, net2: 'Net') -> bool:
        return net2.includes_ip(self.network_address) and self.cidr.value >= net2.cidr.value

    def includes_ip(self, ip: IP) -> bool:
        return (ip.value & self.cidr.mask) == self.network_address.value

    def is_identical(self, net2: 'Net') -> bool:
        return (self.network_address.value == net2.network_address.value and
                self.cidr.value == net2.cidr.value)
